Add support for custom items to editor right-click context menus

citizenll opened the pull request:
### Add right-click menu plugin functionality to implement this [proposal](https://github.com/godotengine/godot-proposals/issues/2401)

Currently, I am only working on 3.x, but in order to incorporate this feature, this is a feature version of 4.x. The difference compared to 3.x is that additional submenu options have been added (currently only expanding the create new option for filesystem).

> submenu options example

```gdscript
extends EditorContextMenuPlugin
const PLUG = preload("res://addons/context-plugin/plug.svg")

func _can_handle(paths: PackedStringArray) -> bool:
	return true

func _init() -> void:
	add_context_menu_item(PLUG, "Svg...", _handle_create_svg, SUBMENU_SLOT_FILESYSTEM_CREATE)

func _handle_create_svg(args):
	print(args)

```
Please refer to the 3.x submission for a more detailed explanation [here](https://github.com/godotengine/godot/pull/94525)

* *Bugsquad edit, closes: https://github.com/godotengine/godot-proposals/issues/2401* Closes https://github.com/godotengine/godot-proposals/issues/8254

citizenll committed:
Add support for custom items to editor right-click context menus

AThousandShips commented on editor/plugins/editor_context_menu_plugin.cpp:
```diff
@@ -0,0 +1,210 @@
+
+void EditorContextMenu::add_context_menu_plugin(ContextMenuSlot slot, const Ref<EditorContextMenuPlugin> &p_plugin) {
+	Ref<EditorContextMenuPlugin> *plugin_list = get_plugin_array(slot);
+	ERR_FAIL_COND(plugin_list[MAX_PLUGINS - 1] != nullptr);
```
```suggestion
	ERR_FAIL_COND(plugin_list[MAX_PLUGINS - 1].is_valid());
```

AThousandShips commented on editor/plugins/editor_context_menu_plugin.cpp:
```diff
@@ -0,0 +1,210 @@
+	Ref<EditorContextMenuPlugin> plugin = p_plugin;
+
+	for (int i = 0; i < MAX_PLUGINS; ++i) {
+		if (plugin_list[i] == nullptr) {
```
```suggestion
		if (plugin_list[i].is_null()) {
```

AThousandShips commented on editor/plugins/editor_context_menu_plugin.cpp:
```diff
@@ -0,0 +1,210 @@
+
+	for (int i = 0; i < MAX_PLUGINS; i++) {
+		Ref<EditorContextMenuPlugin> plugin = plugin_list[i];
+		if (plugin == nullptr) {
```
```suggestion
		if (plugin.is_null()) {
```

AThousandShips commented on editor/plugins/editor_context_menu_plugin.cpp:
```diff
@@ -0,0 +1,210 @@
+
+	for (int i = 0; i < MAX_PLUGINS; i++) {
+		Ref<EditorContextMenuPlugin> plugin = plugin_list[i];
+		if (plugin == nullptr) {
```
```suggestion
		if (plugin.is_null()) {
```

AThousandShips commented on editor/plugins/editor_context_menu_plugin.cpp:
```diff
@@ -0,0 +1,210 @@
+
+	for (int i = 0; i < MAX_PLUGINS; i++) {
+		Ref<EditorContextMenuPlugin> plugin = plugin_list[i];
+		if (plugin == nullptr) {
```
```suggestion
		if (plugin.is_null()) {
```

AThousandShips commented on editor/plugins/editor_plugin.cpp:
```diff
@@ -479,6 +480,16 @@ void EditorPlugin::remove_scene_post_import_plugin(const Ref<EditorScenePostImpo
 }
 
+void EditorPlugin::add_context_menu_plugin(ContextMenuSlot p_slot, const Ref<EditorContextMenuPlugin> &p_plugin) {
+	ERR_FAIL_COND(!p_plugin.is_valid());
```
```suggestion
	ERR_FAIL_COND(p_plugin.is_null());
```

AThousandShips commented on editor/plugins/editor_plugin.cpp:
```diff
@@ -479,6 +480,16 @@ void EditorPlugin::remove_scene_post_import_plugin(const Ref<EditorScenePostImpo
+}
+
+void EditorPlugin::remove_context_menu_plugin(ContextMenuSlot p_slot, const Ref<EditorContextMenuPlugin> &p_plugin) {
+	ERR_FAIL_COND(!p_plugin.is_valid());
```
```suggestion
	ERR_FAIL_COND(p_plugin.is_null());
```

AThousandShips commented on editor/plugins/script_editor_plugin.cpp:
```diff
@@ -3318,6 +3329,17 @@ void ScriptEditor::_make_script_list_context_menu() {
+	Vector<String> selected_paths;
+	if (se) {
+		Ref<Resource> scr = se->get_edited_resource();
+		if (scr != nullptr && scr.is_valid()) {
```
```suggestion
		if (scr.is_valid()) {
```
Unless you meant
```suggestion
		if (scr.is_valid() && scr->is_valid()) {
```

AThousandShips commented on editor/plugins/editor_plugin.h:
```diff
@@ -101,6 +102,12 @@ class EditorPlugin : public Node {
+	enum ContextMenuSlot {
+		CONTEXT_SLOT_SCENE_TREE,
+		CONTEXT_SLOT_FILESYSTEM,
+		CONTEXT_SLOT_SCRIPT_EDITOR
```
```suggestion
		CONTEXT_SLOT_SCRIPT_EDITOR,
```

AThousandShips commented on editor/plugins/editor_context_menu_plugin.h:
```diff
@@ -0,0 +1,112 @@
+	enum ContextMenuSlot {
+		CONTEXT_SLOT_SCENE_TREE,
+		CONTEXT_SLOT_FILESYSTEM,
+		CONTEXT_SLOT_SCRIPT_EDITOR
```
```suggestion
		CONTEXT_SLOT_SCRIPT_EDITOR,
```

AThousandShips commented on editor/plugins/editor_context_menu_plugin.h:
```diff
@@ -0,0 +1,112 @@
+		CONTEXT_ITEM_ID_BASE = 200
+	};
+
+	static void add_context_menu_plugin(ContextMenuSlot slot, const Ref<EditorContextMenuPlugin> &p_plugin);
```
```suggestion
	static void add_context_menu_plugin(ContextMenuSlot p_slot, const Ref<EditorContextMenuPlugin> &p_plugin);
```
For all

AThousandShips commented on editor/plugins/editor_context_menu_plugin.h:
```diff
@@ -0,0 +1,112 @@
+public:
+	enum ContextSubMenuSlot {
+		SUBMENU_SLOT_NONE = 0,
+		SUBMENU_SLOT_FILESYSTEM_CREATE
```
```suggestion
		SUBMENU_SLOT_FILESYSTEM_CREATE,
```

AThousandShips commented on editor/plugins/editor_context_menu_plugin.cpp:
```diff
@@ -0,0 +1,210 @@
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "editor_context_menu_plugin.h"
```
```suggestion
#include "editor_context_menu_plugin.h"

```
Line after header

AThousandShips commented on editor/plugins/editor_context_menu_plugin.cpp:
```diff
@@ -0,0 +1,210 @@
+			break;
+		}
+		int submenu = plugin->get_meta("submenu", 0);
+		if (!plugin->can_handle(p_paths) || submenu) {
```
```suggestion
		if (submenu || !plugin->can_handle(p_paths)) {
```
Cheaper condition first

AThousandShips commented on editor/plugins/editor_context_menu_plugin.cpp:
```diff
@@ -0,0 +1,210 @@
+		}
+	}
+}
+//scene_tree_dock
```
```suggestion

// scene_tree_dock
```
Space between methods

AThousandShips commented on editor/plugins/editor_context_menu_plugin.cpp:
```diff
@@ -0,0 +1,210 @@
+	}
+	invoke_plugin_callback(p_slot, p_option, nodes);
+}
+//script_editor
```
```suggestion

// script_editor
```

AThousandShips commented on editor/plugins/editor_context_menu_plugin.cpp:
```diff
@@ -0,0 +1,210 @@
+void EditorContextMenu::options_pressed(ContextMenuSlot p_slot, int p_option, const Ref<Resource> &p_script) {
+	invoke_plugin_callback(p_slot, p_option, p_script);
+}
+//filesystem_dock
```
```suggestion

// filesystem_dock
```

AThousandShips commented on editor/plugins/editor_context_menu_plugin.cpp:
```diff
@@ -0,0 +1,210 @@
+	}
+}
+
+//Handle submenu item
```
```suggestion
// Handle submenu item.
```

AThousandShips commented on editor/plugins/script_editor_plugin.cpp:
```diff
@@ -1394,6 +1395,16 @@ void ScriptEditor::_menu_option(int p_option) {
 		}
 	}
 
+	// Context menu options
```
```suggestion
	// Context menu options.
```

AThousandShips commented on editor/plugins/script_editor_plugin.cpp:
```diff
@@ -3318,6 +3329,17 @@ void ScriptEditor::_make_script_list_context_menu() {
 	context_menu->set_item_disabled(context_menu->get_item_index(WINDOW_MOVE_DOWN), tab_container->get_current_tab() >= tab_container->get_tab_count() - 1);
 	context_menu->set_item_disabled(context_menu->get_item_index(WINDOW_SORT), tab_container->get_tab_count() <= 1);
 
+	//Context menu plugin
```
```suggestion
	// Context menu plugin.
```

AThousandShips commented on editor/scene_tree_dock.cpp:
```diff
@@ -1473,6 +1474,13 @@ void SceneTreeDock::_tool_selected(int p_tool, bool p_confirm_override) {
 		} break;
 
 		default: {
+			// Editor context plugin
```
```suggestion
			// Editor context plugin.
```

AThousandShips commented on editor/filesystem_dock.cpp:
```diff
@@ -3160,6 +3164,9 @@ void FileSystemDock::_file_and_folders_fill_popup(PopupMenu *p_popup, const Vect
 		new_menu->add_icon_item(get_editor_theme_icon(SNAME("Object")), TTR("Resource..."), FILE_NEW_RESOURCE);
 		new_menu->add_icon_item(get_editor_theme_icon(SNAME("TextFile")), TTR("TextFile..."), FILE_NEW_TEXTFILE);
+
+		//Support plugins in create new...
```
```suggestion
		// Support plugins in create new...
```

AThousandShips commented on editor/plugins/editor_context_menu_plugin.cpp:
```diff
@@ -0,0 +1,210 @@
+	for (int i = 0; i < MAX_PLUGINS; ++i) {
+		if (plugin_list[i] == p_plugin.ptr()) {
+			plugin_list[i].unref();
+			plugin_list[i] = nullptr;
```
```suggestion
```
`unref` already does this

AThousandShips commented on editor/plugins/editor_context_menu_plugin.cpp:
```diff
@@ -0,0 +1,210 @@
+		Array params = plugin->get_meta("parameters");
+		Ref<Texture2D> icon = params[0];
+		int idx = plugin->get_meta("idx");
+		if (!icon.is_null() && icon.is_valid()) {
```
```suggestion
		if (icon.is_valid()) {
```

citizenll replied:
Thank you for pointing it out, all formats have been adjusted.

KoBeWi commented:
So I tested this and it works, but the API is super weird.
`add_context_menu_item()` does not add an item, it defines an item that will appear when plugin is activated. This means that each plugin can add only one menu item. The function's first argument is an icon, which is completely optional for menus. Also if you don't call this method, the editor will crash.
There is `SUBMENU_SLOT_NONE` which does nothing (?) and `SUBMENU_SLOT_FILESYSTEM_CREATE`, which is a part of different enum and can't be used.

I think instead of `_can_handle()`, there could be `_popup_menu(arguments: Array)` virtual function. Though maybe `paths` is fine, not sure where else we will want to add options.
Inside that function the user is supposed to call `add_context_menu_item()`, as many times as they want. I'd change the signature to `add_context_menu_item(text: String, callback: Callable, icon: Texture2D = null, shortcut: Shortcut = null)`. The plugin would have a vector of structs representing added menu items. The editor would then use this vector to add menu items to the PopupMenu.

Instead of keeping an array of plugins per slot, I think it can be a single LocalVector. When opening menu, go through the vector and check the plugin slot. If the slot matches, call `_popup_menu()` and then add options to menu.

To make your life easier, this can be handled with internal methods:
`add_options_from_plugins(PopupMenu *p_menu, ContextMenuSlot p_slot, const PackedStringArray &p_paths)` (this can be a method in EditorData; you need to keep the registered plugins somewhere)
`EditorContextMenuPlugin::add_options(const PackedStringArray &p_paths)` - it would clear the option structs vector, call the virtual `_popup_menu()` and then add items based on added options.

Feel free to ask if something is not clear. You can also join [the developer chat](https://chat.godotengine.org/).

AThousandShips commented on editor/editor_data.cpp:
```diff
@@ -509,6 +511,107 @@ EditorPlugin *EditorData::get_extension_editor_plugin(const StringName &p_class_
+			}
+		}
+	}
+	// invoke submenu items.
```
```suggestion
	// Invoke submenu items.
```

AThousandShips commented on editor/editor_data.cpp:
```diff
@@ -509,6 +511,107 @@ EditorPlugin *EditorData::get_extension_editor_plugin(const StringName &p_class_
+	for (int i = 0; i < context_menu_plugins.size(); i++) {
+		if (context_menu_plugins[i].p_slot != p_slot || context_menu_plugins[i].plugin.is_null()) {
+			continue;
+			;
```
```suggestion
```

AThousandShips commented on editor/editor_data.cpp:
```diff
@@ -509,6 +511,107 @@ EditorPlugin *EditorData::get_extension_editor_plugin(const StringName &p_class_
+		HashMap<String, EditorContextMenuPlugin::ContextMenuItem> &items = plugin->context_menu_items;
+		for (KeyValue<String, EditorContextMenuPlugin::ContextMenuItem> &E : items) {
+			EditorContextMenuPlugin::ContextMenuItem &item = E.value;
+			if (p_option != item.idx || item.callable.is_null()) {
```
```suggestion
			if (p_option != item.idx || !item.callable.is_valid()) {
```
Callable validity should check `is_valid` not `is_null` as they are not complementary

citizenll commented:
Based on the recommendations, the final API has been adjusted; ```_can_handle``` has been changed to ```_popup_menu```. The method ```add_context_menu_item ```will store data in a local HashMap, meaning that menu items with the same name will be unique. Additionally, to enhance understanding, submenu slots are now on the same level as other slots. Below is the example code.

```gdscript
extends EditorContextMenuPlugin
const PLUG = preload("res://addons/csv-update/plug.svg")

func _popup_menu(paths: PackedStringArray) -> bool:
	## The return result will determine whether to display the added options.
	add_context_menu_item("File Custom options", handle, PLUG)
	return true

func handle(paths):
	print("file context item handle:", paths)

```

citizenll commented on editor/plugins/editor_context_menu_plugin.h:
```diff
@@ -0,0 +1,71 @@
+public:
+	virtual bool add_options(const Vector<String> &p_paths);
+	void add_context_menu_item(const String &p_name, const Callable &p_callable, const Ref<Texture2D> &p_texture, const Ref<Shortcut> &p_shortcut);
+};
```
Thank you for pointing that out; your response speed is truly impressive.👋

AThousandShips commented on editor/editor_data.cpp:
```diff
@@ -509,6 +511,106 @@ EditorPlugin *EditorData::get_extension_editor_plugin(const StringName &p_class_
+	bool add_separator = false;
+
+	for (int i = 0; i < context_menu_plugins.size(); i++) {
+		ContextMenu &cm = context_menu_plugins.get(i);
```
```suggestion
		ContextMenu &cm = context_menu_plugins.get(i);
```
```suggestion
		const ContextMenu &cm = context_menu_plugins[i];
```

KoBeWi commented on editor/editor_data.cpp:
```diff
@@ -509,6 +511,104 @@ EditorPlugin *EditorData::get_extension_editor_plugin(const StringName &p_class_
+		}
+		if (!add_separator) {
+			add_separator = true;
+			p_popup->add_separator();
```
This should only happen after the condition below. Otherwise the menu can get an empty separator.
![image](https://github.com/user-attachments/assets/26e7ad45-dea8-4b71-8c49-697487f36aee)

KoBeWi commented:
After the changes the API and implementation look much, but it needs some fixes. Note that you'll need to add documentation to new class and methods. Run Godot executable with `--doctool` and fill the descriptions.

- [ ] Context menus don't fit custom icons. You'll need to use `set_item_icon_max_width()`.
![image](https://github.com/user-attachments/assets/7cf16e87-7310-4be9-b0e7-7c160763f75a)

- [ ] You are not clearing previously added menu options. If you add an option from plugin and then rename it in the code, the plugin will add both new and old options.

- [ ] I think the `bool` return value from `_popup_menu()` is unnecessary. If you resolve the above point, just check if any options were added in the callback.

 - [ ] Shortcuts don't work correctly, they are added as another, empty, option.
![image](https://github.com/user-attachments/assets/71fa4cd3-ad90-465f-87ee-c73e0295113f)

KoBeWi commented on editor/plugins/editor_plugin.h:
```diff
@@ -101,6 +102,13 @@ class EditorPlugin : public Node {
 		AFTER_GUI_INPUT_CUSTOM,
 	};
 
+	enum ContextMenuSlot {
```
This could be moved to EditorContextMenuPlugin instead.

KoBeWi commented on editor/plugins/editor_context_menu_plugin.cpp:
```diff
@@ -0,0 +1,62 @@
+	if (context_menu_items.has(p_name)) {
+		context_menu_items.erase(p_name);
+	}
+	item.idx = EditorData::CONTEXT_MENU_ITEM_ID_BASE + start_idx + context_menu_items.size() + 1;
```
What's +1 for?

KoBeWi commented on editor/editor_data.h:
```diff
@@ -122,8 +124,27 @@ class EditorData {
+	};
+
+	enum {
+		CONTEXT_MENU_ITEM_ID_BASE = 200
```
I'd make it 1000, just to be safe.
This can also be `inline static constexpr int`.

KoBeWi commented:
Looking at all the new methods in EditorData, I think they can be moved to some new class called EditorContextMenuManager 🤔 Like EditorInspector manages inspector plugins or 3D editor manages gizmo plugins etc.
Though they are completely internal, so this can be done in a future refactor.

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -3160,6 +3164,9 @@ void FileSystemDock::_file_and_folders_fill_popup(PopupMenu *p_popup, const Vect
 		new_menu->add_icon_item(get_editor_theme_icon(SNAME("Object")), TTR("Resource..."), FILE_NEW_RESOURCE);
 		new_menu->add_icon_item(get_editor_theme_icon(SNAME("TextFile")), TTR("TextFile..."), FILE_NEW_TEXTFILE);
+
+		// Support plugins in create new...
```
Such comments are unnecessary. The line is self-explanatory.

KoBeWi commented on editor/editor_data.h:
```diff
@@ -176,6 +197,20 @@ class EditorData {
+	void remove_context_menu_plugin(ContextMenuSlot slot, const Ref<EditorContextMenuPlugin> &p_plugin);
+	//
+	void add_options_from_plugins(PopupMenu *p_popup, ContextMenuSlot p_slot, const Vector<String> &p_paths);
+	// for filesystem dock.
```
Instead of commenting, you could rename the methods. They don't need to use the same name.

citizenll replied:
Earlier, I thought that if we clear after adding options, the subsequent event responses wouldn't be able to find the functions. Now, I believe this clearing should happen before adding options to the plugin. The other suggestions are very useful, and the +1 for starting the ID from 1 is not essential. Thank you! 👋~

KoBeWi commented on editor/editor_data.cpp:
```diff
@@ -509,6 +511,103 @@ EditorPlugin *EditorData::get_extension_editor_plugin(const StringName &p_class_
+
+			if (item.icon.is_valid()) {
+				p_popup->add_icon_item(item.icon, item.item_name, item.idx);
+				p_popup->set_item_icon_max_width(-1, 32);
```
This should use `class_icon_size` theme constant instead of hard-coded `32`.

KoBeWi commented on editor/editor_data.h:
```diff
@@ -176,6 +195,17 @@ class EditorData {
 	bool has_extension_editor_plugin(const StringName &p_class_name);
 	EditorPlugin *get_extension_editor_plugin(const StringName &p_class_name);
 
+	// context menu plugin.
```
```suggestion
	// Context menu plugin.
```

KoBeWi commented on editor/editor_data.h:
```diff
@@ -176,6 +195,17 @@ class EditorData {
+	// context menu plugin.
+	void add_context_menu_plugin(ContextMenuSlot slot, const Ref<EditorContextMenuPlugin> &p_plugin);
+	void remove_context_menu_plugin(ContextMenuSlot slot, const Ref<EditorContextMenuPlugin> &p_plugin);
+	//
```
?

KoBeWi commented on editor/plugins/editor_context_menu_plugin.cpp:
```diff
@@ -0,0 +1,61 @@
+
+#include "core/input/shortcut.h"
+#include "scene/resources/texture.h"
+#include <editor/editor_node.h>
```
```suggestion
#include "editor/editor_node.h"
```

KoBeWi commented on editor/plugins/editor_context_menu_plugin.cpp:
```diff
@@ -0,0 +1,61 @@
+}
+
+void EditorContextMenuPlugin::add_context_menu_item(const String &p_name, const Callable &p_callable, const Ref<Texture2D> &p_texture, const Ref<Shortcut> &p_shortcut) {
+	ERR_FAIL_COND(context_menu_items.size() == MAX_ITEMS);
```
Could also return error if the given name already exists.

KoBeWi commented on editor/plugins/editor_context_menu_plugin.h:
```diff
@@ -0,0 +1,67 @@
+	inline static constexpr int MAX_ITEMS = 100;
+
+	struct ContextMenuItem {
+		int idx;
```
```suggestion
		int idx = 0;
```

KoBeWi commented:
https://github.com/godotengine/godot/pull/94582#discussion_r1701832562 should be done now, because afterwards it's a breaking change. You also still need to fill documentation descriptions.

I noticed that shortcuts are not working, but this will require more changes. Seems like `activate_item_by_event()` is not called automatically for context menus and has to be manually handled. It can be fixed in a follow-up.

Context menu callbacks receive different arguments than `_popup_menu()`. E.g. in scene tree `_popup_menu()` gets node paths, but once the item is used, the callback receives nodes. Not sure if this is intuitive 🤔 Although it makes it easier to access the actual objects instead of their paths.

citizenll commented:
I'm somewhat confused about the shortcut keys. I'm not sure if the function for setting shortcut keys in PopupMenu only displays them as text options or if it actually can link to shortcut key actions. If the goal is to implement a feature similar to a global response to shortcut keys, it feels like it's addressing a different function—specifically, "binding global shortcut keys in plugins." Currently, what I can do is just use PopupMenu->add_shortcut. 😭

KoBeWi commented:
There is `activate_item_by_event()`, which is called in MenuButton, OptionButton and MenuBar. Basically you have to pass shortcuts manually. I think it's not done for context menus, because the list of items depends on context, so not all shortcuts will always be functional.

The way it works in the editor is that all possible shortcuts are defined beforehand and then they are bound to items when the menu is refreshed. Then they are handled manually, independently from the menu. See here for example:

https://github.com/godotengine/godot/blob/3b3d6226c5cd3f9897160815990e2e234f165605/editor/scene_tree_dock.cpp#L171-L178

Not sure how to best solve it. Ideally the shortcuts should work like in the editor, but also be optional. Though I wonder if we even  need shortcuts for custom options.

citizenll commented:
I went to test the function `activate_item_by_event`. I added the execution of `activate_item_by_event(p_event, true)` here https://github.com/godotengine/godot/blob/568589c9d8c763bfb3a4348174d53b42d7c59f21/editor/filesystem_dock.cpp#L3549 and found an issue through debugging: when the menu has not been created, the event responds from here, but at this point, there is no context menu created, so our event cannot be bound. Moreover, once the context menu is created, it seems the focus is on the menu, and the event does not respond again. It feels like going back to creating global shortcuts for the plugin and needing to register some information in advance when there’s no context menu created.

KoBeWi commented:
Well yeah, that's what I said in my previous comment. PopupMenu's items are usually dynamic, so not all shortcuts are available at all times.

citizenll commented:
I have added a test version, which currently only affects the filesystem. You can check if this format is suitable. I also added a new API method: `void set_item_shortcut(const Ref<Shortcut> &p_shortcut, const Callable &p_callable);`

```gdscript
extends EditorContextMenuPlugin
const PLUG = preload("res://addons/csv-update/plug.svg")
const SHORTCUT = preload("res://test.tres")

func _init() -> void:
	set_item_shortcut(SHORTCUT, handle);

func _popup_menu(paths: PackedStringArray):
	add_context_menu_item("File Custom options", handle, PLUG)


func handle(paths):
	print("file context item handle:", paths)
```

KoBeWi commented:
I assume the implementation is temporary? The plugin can add multiple options, so the method should allow adding multiple shortcuts. I'd rename the method to `add_menu_shortcut()`.

The shortcut itself works correctly.

citizenll commented:
![image](https://github.com/user-attachments/assets/6010594f-0ea5-4bc5-ad1c-8852ea5b6cc9)
The scene tree and script editor plugin both receive input events simultaneously... How should I handle this...? When I use the same shortcut, they both respond at the same time, which means that the shortcuts in these two areas cannot be the same.
![image](https://github.com/user-attachments/assets/9f5ae35d-4c20-4158-b7c3-554d726bdbf0)

The code supporting shortcuts has been updated. Currently, it should restrict the scene tree and code editor from using the same shortcuts. If there are no other issues, I will supplement the documentation, and we can close this PR.

KoBeWi commented on doc/classes/EditorContextMenuPlugin.xml:
```diff
@@ -0,0 +1,49 @@
+				To register the shortcut for the context menu, call this function within the [method Object._init] function, even if the context menu has not been created yet.
+				Note that this method should only be invoked from [method Object._init]; otherwise, the shortcut will not be registered correctly.
+				[codeblock]
+				func _init() -> void:
```
```suggestion
				func _init() -&gt; void:
```
Although we don't put type hints in the docs.

KoBeWi commented on editor/editor_data.cpp:
```diff
@@ -509,6 +512,140 @@ EditorPlugin *EditorData::get_extension_editor_plugin(const StringName &p_class_
+}
+
+int EditorData::match_context_menu_shortcut(ContextMenuSlot p_slot, const Ref<InputEvent> &p_event) {
+	for (int i = 0; i < context_menu_plugins.size(); i++) {
+		const ContextMenu &cm = context_menu_plugins[i];
```
```suggestion
	for (const ContextMenu &cm : context_menu_plugins) {
```

KoBeWi commented on doc/classes/EditorPlugin.xml:
```diff
@@ -621,6 +630,13 @@
+		<method name="remove_context_menu_plugin">
+			<return type="void" />
+			<param index="0" name="slot" type="int" enum="EditorPlugin.ContextMenuSlot" />
+			<param index="1" name="plugin" type="EditorContextMenuPlugin" />
```
Needs description. Same below.

KoBeWi commented on doc/classes/EditorPlugin.xml:
```diff
@@ -864,5 +880,13 @@
 			Pass the [InputEvent] to other editor plugins except the main [Node3D] one. This can be used to prevent node selection changes and work with sub-gizmos instead.
 		</constant>
+		<constant name="CONTEXT_SLOT_SCENE_TREE" value="0" enum="ContextMenuSlot">
+		</constant>
```
Also needs description.

KoBeWi commented on editor/editor_data.cpp:
```diff
@@ -509,6 +512,140 @@ EditorPlugin *EditorData::get_extension_editor_plugin(const StringName &p_class_
+void EditorData::add_options_from_plugins(PopupMenu *p_popup, ContextMenuSlot p_slot, const Vector<String> &p_paths) {
+	bool add_separator = false;
+
+	for (int i = 0; i < context_menu_plugins.size(); i++) {
+		const ContextMenu &cm = context_menu_plugins[i];
```
```suggestion
	for (const ContextMenu &cm : context_menu_plugins) {
```

KoBeWi commented on editor/plugins/editor_context_menu_plugin.h:
```diff
@@ -0,0 +1,69 @@
+#ifndef EDITOR_CONTEXT_MENU_PLUGIN_H
+#define EDITOR_CONTEXT_MENU_PLUGIN_H
+
+#include "core/io/resource.h"
```
This should be `ref_counted.h`.

KoBeWi reviewed (approved):
I think functionality-wise it's ok now and the implementation is good enough. It can be merged once you address some last comments.

AThousandShips commented on doc/classes/EditorContextMenuPlugin.xml:
```diff
@@ -0,0 +1,49 @@
+				Add custom options to the context menu of the currently specified slot.
+				To trigger a [param shortcut] before the context menu is created, please additionally call the [method add_menu_shortcut] function.
+				[codeblock]
+				func _popup_menu(paths: PackedStringArray):
```
```suggestion
				func _popup_menu(paths):
```
We [don't add](https://docs.godotengine.org/en/stable/contributing/documentation/docs_writing_guidelines.html#dynamic-vs-static-typing) static typing to the documentation

AThousandShips commented on editor/editor_data.cpp:
```diff
@@ -509,6 +512,138 @@ EditorPlugin *EditorData::get_extension_editor_plugin(const StringName &p_class_
+		int shortcut_base_idx = EditorData::CONTEXT_MENU_ITEM_ID_BASE + plugin->start_idx;
+		for (KeyValue<Ref<Shortcut>, Callable> &E : plugin->context_menu_shortcuts) {
+			if (shortcut_base_idx + shortcut_idx == p_option) {
+				const Callable &p_callable = E.value;
```
```suggestion
				const Callable &callable = E.value;
```
Don't use `p_` for non-arguments

citizenll replied:
Great,  done~👋

akien-mga commented:
Amended the commit message to match our preferred style for clarity.

citizenll commented:
Oh, what have I done? I thought I was just supposed to modify the commit message. Did my resubmission break anything?

AThousandShips commented:
You fixed the message but it had already been fixed, and your branch was behind this one so you pulled it back to an earlier version

citizenll commented:
Will my new force push affect this modification? Do I need to correct this error, or can I leave it as is? I'm not sure how to handle the pull request in this situation.

AThousandShips commented:
You can leave it, I updated the PR message to match, but it's good in the future to pay attention to what happens in your PR to avoid undoing or redoing things

akien-mga commented:
My bad, I should have been more explicit that I did the action myself, and wasn't asking for you to do it :)
